README.md
[![Build Status](https://travis-ci.org/mr5/counter-rank.svg)](https://travis-ci.org/mr5/counter-rank)
[![Still Maintained](http://stillmaintained.com/mr5/counter-rank.png)](http://stillmaintained.com/mr5/counter-rank)
[![License](https://poser.pugx.org/mr5/counter-rank/license.svg)](https://packagist.org/packages/mr5/counter-rank)

## 简介
Counter-Rank 是一个使用 PHP 语言编写的、基于 Redis 的排名与计数类库，主要是对 Redis 的 zset 操作进行了封装。并提供了一个 [`JSClientHandler`](lib/mr5/CounterRank/JSClientHandler.php) 类，可以快速地生成 JsonP 接口用于满足一些静态页面的计数需求以及将数字渲染到页面。

本类库的源代码中有详细的注释，你可以使用类 PHPDoc 的工具生成 classes reference manual 。

## 安装
在 `composer.json` 中添加：
```json
{
	require: "mr5/counter-rank"
}
```
然后安装
```shell
composer install
```
## 注意事项
打开 `phpunit.xml` 文件，找到以下片段：
```xml
<php>
	<!-- Redis -->
	<const name="REDIS_SERVER_HOST" value="127.0.0.1" />
	<const name="REDIS_SERVER_PORT" value="6379" />
	<const name="REDIS_NAMESPACE" value="__unitTest___" />
</php>
```
如你要运行本类库的单元测试，请注意其中 `REDIS_NAMESPACE` 定义的命名空间，以防误删你的 Redis 数据。如有必要，可以修改该值。 
## 基础使用
```php
use mr5\CounterRank\CounterRank;

// 命名空间 namespace 用于区分不同的项目。
// 分组名是一类 items 的分组。
// 比如要统计的是文章，则分组名可以是 articles，评论的分组名可以是 comments。
$counterRank = new CounterRank(
	'redis_host', 
	'redis_port', 
	'namespace', 
	'分组名'
);

// 创建一个item，create 方法可以接收一个数字作为默认值，留空则为0。
// 下面的`900310`可以看做是文章 ID、评论 ID 等等。
$counterRank->create('900310', 0);

// 删除一个分组，`articles` 是分组名
$counterRank->deleteGroup('articles');

// 删除一个 item
$counterRank->delete('900310');

// 递增指定键名的值，如为负数则为递减。
// 这里对 `900310` 这篇文章递增了 1
$counterRank->increase('900310', 1);
// 递减
$counterRank->increase('900310', -1);

// 倒序排序，最多 10 个。
$counterRank->rank(10, 'desc');
// 正序排序，最多 10 个
$counterRank->rank(10, 'asc');
// 最高的 10 个
$counterRank->top10();
// 最低的 10 个
$counterRank->down10();

// 设置当操作一个不存在的 keys 时的处理闭包。
// 该闭包将接收两个参数，第一个参数是 key ，
// 第二个参数是当前 CounterRank 对象。
// 如修复后该 key 可以操作时返回 true，否则返回 false。
// 可以用于自动创建 item
$counterRank->setFixMiss(function($key, CounterRank $counterRank) {
            return $counterRank->create($key, 0) > 0;
});


```
## JSClientHandler 的使用
`JSClientHandler` 是一个用于生成 JS 客户端的工具类。
它可以实现以下功能：
* token 验证
* get: 读取指定分组指定键的数值;
* increase: 递增;
* rank: 获取排名;
* 指定 callback ，以实现 JsonP

以下是一个自定义控制器参考，请根据自己使用的框架以及需求进行修改，另外还建议阅读 [`JSClientHandler.php`](lib/mr5/CounterRank/JSClientHandler.php) 的源代码，内含详细的 PHP Doc：

```php
use mr5\CounterRank\JSClientHandler;

/**
 *
 * JS 客户端控制器示例，请根据自己使用的框架以及需求进行修改
 *
 */
class ExampleController  extends Controller
{
	// 配置客户端的 token，键是分组名，值是 token 值。每个分组都必须指定，未指定的则不允许通过 JS Client 访问
    protected $tokens = array(
        'articles' => '1234567890JQK',
        'comments' => 'abcdefghijk'
    );
    protected $redis_host = 'localhost';
    protected $redis_port = 6379;

	// 命名空间，建议使用项目名称
    protected $namespace = 'project_name';
	// 递增的步长，如设置成负数，则 increase 方法执行的是递减
    protected $increaseStepSize = 1;

    // @var JSClientHandler|null
    protected $handlerInstance = null;

    public function __construct()
    {
        $this->handlerInstance = new JSClientHandler(
						$this->redis_host, 
						$this->redis_port, 
						$this->namespace, 
						$this->tokens, 
						$this->increaseStepSize
		);
    }

    // 读取
    public function getAction()
    {
        $this->handlerInstance->handleGet(
			$_REQUEST['token'], 
			$_REQUEST['group'], 
			$_REQUEST['keys'], 
			$_REQUEST['callback']
		);
    }

    // 递增
    public function increaseAction()
    {
        $this->handlerInstance->handleIncrease(
			$_REQUEST['token'], 
			$_REQUEST['group'], 
			$_REQUEST['keys'], 
			$_REQUEST['callback']
		);
    }

    // 排名
    public function rankAction()
    {
        $this->handlerInstance->handleRank(
			$_REQUEST['token'], 
			$_REQUEST['group'], 
			$_REQUEST['type'], 
			$_REQUEST['limit'], 
			$_REQUEST['callback']
		);
    }

    // 最高的十个数据
    public function top10Action()
    {
        $this->handlerInstance->handleTop10(
			$_REQUEST['token'], 
			$_REQUEST['group'], 
			$_REQUEST['callback']
		);

    }

    // 最低的十个数据
    public function down10Action()
    {
        $this->handlerInstance->handleDown10(
			$_REQUEST['token'], 
			$_REQUEST['group'], 
			$_REQUEST['callback']
		);

    }
}

```
### 自定义 token 验证规则

token 默认验证方式是判断用户递交的 token 是否等于约定的 token ，
你可以通过传递一个闭包给 `setTokenVerifier` 方法来自定义这个验证规则。
下面是自定义 tokenVerifier 的例子：
```php

$jsClientHandler->setTokenVerifier(function (
			$operation,	// 操作名
			$userToken, 	// 客户端提交的 token
			$token, 		// 服务器端约定的 token
			$group, 		// 分组名
			$keys) {
                $str = $token.$group;
                if($keys) {
                    $str .= $keys;
                }
                return md5($str) == $userToken;
            });
```

## 持久化计数数据

 [`CounterRank`](lib/mr5/CounterRank/CounterRank) 类提供了一个名为 `persistHelper` 方法来帮助你持久化计数器中的数据，如同步到 MySQL 数据库。`persistHelper` 方法接收两个参数，第一个参数是一个闭包，该闭包包含了你自定义的同步逻辑，闭包的参数是 items 键值对数组。第二个参数有三个选项：
 
 * `CounterRank::PERSIST_WITH_DELETING`  持久化后的 items 删除，使用场景为使用 CounterRank 统计全量数据。一般需要配合 `$counterRank->setMissFix()` 来初始化不存在的键。
 * `CounterRank::PERSIST_WITH_CLEARING`  持久化后的 items 清零，使用场景为使用 CounterRank 统计增量数据，在比如 MySQL 之类的持久化数据库系统中保留全量数据。
 * `CounterRank::PERSIST_WITH_NOTHING`   持久化后不执行任何操作，使用场景为使用 redis 统计全量数据，并且作为主要持久化途径，其他持久化途径仅作为备份。

下面是一个同步到数据库的例子：

```php
// 注意：本操作可能耗时很长，请在命令行下执行并根据实际情况加入操作系统计划任务，而不是通过 web 请求执行

$counterRank = new CounterRank(
	'redis_host', 
	'redis_port', 
	'namespace', 
	'分组名'
);
$conn = mysql_connect('localhost', 'username', 'password');
mysql_select_db('dbname', $conn);
mysql_set_charset('utf8', $conn);

$counterRank->persistHelper(function($items) {
    $values = '';
    foreach($items AS $post_id=>$heat) {
        if($values != '') {
            $values .= ',';
        }
        $values .= "({$post_id}, {$heat})";
    }
    mysql_query(
        "INSERT INTO `posts`(post_id, heat) VALUES  {$values} ON DUPLICATE KEY UPDATE heat=values(heat);",
        $conn
    );
});
```


CHANGELOG.md
## Change Log

### 0.02 (2014/07/15 10:46 +00:00)
- [a4614f8](https://github.com/mr5/counter-rank/commit/a4614f8116e13cf82ffe8cdc477e9f734bed3676) 增加持久化帮助方法 (@mr5)

### 0.01 (2014/07/15 02:52 +00:00)
- [1b1c9da](https://github.com/mr5/counter-rank/commit/1b1c9da4d367a096fa8a4cf090f81de9f0ab8d00) initialize commit (@mr5)
- [48be1ee](https://github.com/mr5/counter-rank/commit/48be1ee881616486a78044b0bc32a85d03c54b1d) update introduction (@mr5)
- [3445d0d](https://github.com/mr5/counter-rank/commit/3445d0d53c9cecaf296f5df38699f93f7289dcfc) update introduction (@mr5)
- [62f3118](https://github.com/mr5/counter-rank/commit/62f311806ea912fa7af70c41767eb9b54137d61b) update introduction (@mr5)
- [930f2ba](https://github.com/mr5/counter-rank/commit/930f2bae02c5bc9f4f00370f49f4bd79519ae8eb) bug fixed (@mr5)
- [d537e50](https://github.com/mr5/counter-rank/commit/d537e503719151fa2a1dae1c58eb92417cd5fc72) add key exist check (@mr5)
- [8bf6544](https://github.com/mr5/counter-rank/commit/8bf6544a05bcf6265211909c2ebbf1b8629ed188) add key exist check (@mr5)
- [d4d299d](https://github.com/mr5/counter-rank/commit/d4d299d756ba5470dbe85998b841d95b68ff8255) fixed returns of rank when asc type used (@mr5)
- [6b8ae55](https://github.com/mr5/counter-rank/commit/6b8ae55a08ac97898bd5bafd57261d19b72ba910) fixed array_reverse (@mr5)
- [39cb397](https://github.com/mr5/counter-rank/commit/39cb397b5ab574d8d7f52e68ae3a692b4c78d5fc) push permission test (@mr5)
- [5066595](https://github.com/mr5/counter-rank/commit/50665954b6d61cb5cad825730f00d3d2eea412ef) 支持指定一个闭包处理当操作一个不存在的 key (@mr5)
- [7eb3529](https://github.com/mr5/counter-rank/commit/7eb3529c3e9533fa63dcb35817a6a834665ea937) 更新说明 (@mr5)
- [f051fcb](https://github.com/mr5/counter-rank/commit/f051fcb62fa585c0731a8ab4769250014f2bc9e9) update introduction (@mr5)
- [b2a51ab](https://github.com/mr5/counter-rank/commit/b2a51abe9470610183a7f4171bf39243b564adda) update introduction (@mr5)
- [0ff4c51](https://github.com/mr5/counter-rank/commit/0ff4c5138494bd706602964a75b1749f35ebc19c) 增加自定义 token 验证器 (@mr5)

tests/bootstrap.php
<?php
// +----------------------------------------------------------------------
// | [counter-rank]
// +----------------------------------------------------------------------
// | Author: Mr.5 <mr5.simple@gmail.com>
// +----------------------------------------------------------------------
// + Datetime: 14-7-4 下午5:45
// +----------------------------------------------------------------------
// + bootstrap.php
// +----------------------------------------------------------------------
require __DIR__ .'/../vendor/autoload.php';

tests/mr5/CounterRank/BenchmarkTest.php
<?php

namespace mr5\CounterRank;

// +----------------------------------------------------------------------
// | [counter-rank]
// +----------------------------------------------------------------------
// | Author: Mr.5 <mr5.simple@gmail.com>
// +----------------------------------------------------------------------
// + Datetime: 14-7-9 13:30
// +----------------------------------------------------------------------
// + BenchmarkTest.php 性能基准测试
// +----------------------------------------------------------------------

class BenchmarkTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @var \mr5\counterRank\CounterRank
     */
    private $counterRank = null;
    private $groupName = 'CounterAndRankTest';

    protected function setUp()
    {
        $this->counterRank = new CounterRank(REDIS_SERVER_HOST, REDIS_SERVER_PORT, REDIS_NAMESPACE, $this->groupName);
    }
    public function testBenchmark()
    {
        set_time_limit(6000);
        $startTime = microtime(true);

        for($i=1; $i<50000; $i++) {
            $this->counterRank->create('item'.$i, 0);
        }
        $this->assertLessThan(0.9, microtime(true) - $startTime);
    }
    protected function tearDown()
    {
        $this->counterRank->deleteGroup($this->groupName);
    }
}

tests/mr5/CounterRank/CounterRankTest.php
<?php

namespace mr5\CounterRank;

// +----------------------------------------------------------------------
// | [counter-rank]
// +----------------------------------------------------------------------
// | Author: Mr.5 <mr5.simple@gmail.com>
// +----------------------------------------------------------------------
// + Datetime: 14-7-4 下午2:50
// +----------------------------------------------------------------------
// + CounterAndRankTest.php
// +----------------------------------------------------------------------


class CounterRankTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @var \mr5\counterRank\CounterRank
     */
    public  $counterRank = null;
    /**
     * @var string
     */
    private $groupName = 'CounterAndRankTest';
    /**
     * @var array
     */
    private $testData = array();

    protected function setUp()
    {

        $this->counterRank = new CounterRank(REDIS_SERVER_HOST, REDIS_SERVER_PORT, REDIS_NAMESPACE, $this->groupName);

        $this->testData['items'] = array();

        $_numbers = array();
        for ($i = 0; $i < 1000; $i++) {
            $_numbers[] = $i;
        }
        shuffle($_numbers);
        foreach ($_numbers as $_num) {
            $this->testData['items']['testSortItem_' . $_num] = (string)$_num;
        }
        $this->testData['top10'] = array(
            "testSortItem_999" => "999",
            "testSortItem_998" => "998",
            "testSortItem_997" => "997",
            "testSortItem_996" => "996",
            "testSortItem_995" => "995",
            "testSortItem_994" => "994",
            "testSortItem_993" => "993",
            "testSortItem_992" => "992",
            "testSortItem_991" => "991",
            "testSortItem_990" => "990"
        );
        $this->testData['down10'] = array(
            "testSortItem_0" => "0",
            "testSortItem_1" => "1",
            "testSortItem_2" => "2",
            "testSortItem_3" => "3",
            "testSortItem_4" => "4",
            "testSortItem_5" => "5",
            "testSortItem_6" => "6",
            "testSortItem_7" => "7",
            "testSortItem_8" => "8",
            "testSortItem_9" => "9"
        );
    }

    /**
     * 测试删除分组
     */
    public function testDeleteGroup()
    {

        $this->assertEquals(1, $this->counterRank->create('testDeleteGroup', 1));

        $this->assertEquals(2, $this->counterRank->mCreate(array(
            'testDeleteGroup2' => 2,
            'testDeleteGroup3' => 3
        )));

        $this->assertEquals(1, $this->counterRank->get('testDeleteGroup'));
        $this->assertEquals(2, $this->counterRank->get('testDeleteGroup2'));
        $this->assertEquals(3, $this->counterRank->get('testDeleteGroup3'));

        $this->assertTrue($this->counterRank->deleteGroup($this->groupName));

        $this->assertFalse($this->counterRank->get('testDeleteGroup'));
        $this->assertFalse($this->counterRank->get('testDeleteGroup2'));
        $this->assertFalse($this->counterRank->get('testDeleteGroup3'));
    }

    /**
     * 创建、删除和获取测试
     */
    public function testCreateDeleteAndRead()
    {
        $items = array();
        for ($i = 1; $i <= 10; $i++) {
            $items['item' . $i] = $i;
        }
        $keys = array_keys($items);
        $this->assertEquals(1, $this->counterRank->create('testSingleCreate', 1));
        $this->assertEquals(1, $this->counterRank->get('testSingleCreate'));
        $this->assertEquals(1, $this->counterRank->delete('testSingleCreate'));
        $this->assertFalse($this->counterRank->get('testSingleCreate'));

        // 批量插入测试
        $this->assertTrue($this->counterRank->mCreate($items) > 0);
        // 批量获取测试
        $this->assertCount(10, $this->counterRank->mGet($keys));
        $this->assertEquals(3, $this->counterRank->get('item3'));
        // 批量删除
        $this->assertEquals(10, $this->counterRank->mDelete($keys));
        // 批量删除过后应该取不到值
        $this->assertCount(0, $this->counterRank->mGet($keys));


        // 不存在的 key
        $this->assertFalse($this->counterRank->get('keyNotExist'));
    }

    /**
     * 递增测试
     */
    public function testIncrease()
    {
        // 单个元素的递增测试
        $this->assertEquals(1, $this->counterRank->create('testIncrease', 0));
        $this->assertEquals(1, $this->counterRank->increase('testIncrease', 1));
        $this->assertEquals(1, $this->counterRank->get('testIncrease'));
        $this->assertEquals(0, $this->counterRank->increase('testIncrease', -1));
        $this->assertEquals(0, $this->counterRank->get('testIncrease'));
    }

    /**
     * 排序测试
     */
    public function testRank()
    {

        $this->assertEquals(count($this->testData['items']), $this->counterRank->mCreate($this->testData['items']));

        $this->assertEquals($this->testData['top10'], $this->counterRank->rank(10, 'desc'));
        $this->assertNotEquals($this->testData['top10'], $this->counterRank->rank(10, 'asc'));
        $this->assertEquals($this->testData['down10'], $this->counterRank->rank(10, 'asc'));
        $this->assertEquals($this->testData['down10'], $this->counterRank->down10());
        $this->assertEquals($this->testData['top10'], $this->counterRank->top10());
    }

    /**
     * 持久化帮助方法测试，持久化后删除
     */
    public function testPersistHelperWithDeleting()
    {
        $self = $this;

        $itemsSource = $this->testData['items'];
        $this->assertEquals(count($this->testData['items']), $this->counterRank->mCreate($itemsSource));
        $testItems = array();

        $this->counterRank->persistHelper(function (array $_items) use (& $testItems, $self) {
            $testItems = array_merge($testItems, $_items);


            $self->assertEmpty($self->counterRank->mGet(array_keys($_items)));

            unset($_items);
        }, CounterRank::PERSIST_WITH_DELETING);

        arsort($itemsSource);
        $this->assertEquals(json_encode($itemsSource) , json_encode($testItems));
        unset($fileContentExcepted);
        unset($itemsSource);
    }
    /**
     * 持久化帮助方法测试，持久化后不执行任何操作
     */
    public function testPersistHelperWithNothing()
    {
        $self = $this;

        $itemsSource = $this->testData['items'];
        $this->assertEquals(count($this->testData['items']), $this->counterRank->mCreate($itemsSource));
        $testItems = array();
        $this->counterRank->persistHelper(function (array $_items) use (& $testItems, $self) {
            $testItems = array_merge($testItems, $_items);

            $self->assertEquals($_items, $self->counterRank->mGet(array_keys($_items)));

            unset($_items);
        }, CounterRank::PERSIST_WITH_NOTHING);

        arsort($itemsSource);


        $this->assertEquals(json_encode($itemsSource) , json_encode($testItems));
        unset($itemsSource);
        unset($testItems);
    }
    /**
     * 持久化帮助方法测试，持久化后清零
     */
    public function testPersistHelperWithClearing()
    {
        $self = $this;
        $itemsSource = $this->testData['items'];
        $this->assertEquals(count($this->testData['items']), $this->counterRank->mCreate($itemsSource));
        $testItems = array();

        $this->counterRank->persistHelper(function (array $_items) use (& $testItems, $self) {
            $testItems = array_merge($testItems, $_items);

            $itemsExcepted = array();
            foreach($_items AS $key=>$_v) {
                $itemsExcepted[$key] = 0;
            }

            $self->assertEquals($itemsExcepted, $self->counterRank->mGet(array_keys($_items)));

            unset($_items);
            unset($itemsExcepted);
        }, CounterRank::PERSIST_WITH_CLEARING);

        arsort($itemsSource);
        unset($itemsSource['testSortItem_0']);
        $this->assertEquals(json_encode($itemsSource) , json_encode($testItems));
        unset($fileContentExcepted);
        unset($testItems);
    }


    /**
     * 测试 fixMiss 闭包
     */
    public function testFixMiss()
    {
        $this->assertNull($this->counterRank->increase('inexistense', 10));

        $this->counterRank->setFixMiss(function ($key, CounterRank $counterRank) {
            return $counterRank->create($key, 1000) > 0;
        });

        $this->assertEquals(1010, $this->counterRank->increase('inexistense', 10));

    }

    protected function tearDown()
    {
        $this->counterRank->deleteGroup($this->groupName);
    }

}

lib/mr5/CounterRank/JSClientHandler.php
<?php

namespace mr5\CounterRank;

// +----------------------------------------------------------------------
// | [counter-rank]
// +----------------------------------------------------------------------
// | Author: Mr.5 <mr5.simple@gmail.com>
// +----------------------------------------------------------------------
// + Datetime: 14-7-7 下午3:33
// +----------------------------------------------------------------------
// + JSClientHandler.php  JS 客户端处理
// +----------------------------------------------------------------------

/**
 *
 * JS 客户端 handler
 * @package mr5\CounterRank
 *
 * @example
 * ``
 * ```php
 * // 控制器参考
 * class ExampleController  extends Controller
 * {
 * protected $token = array(
 * 'articles' => '1234567890JQK',
 * 'comments' => 'abcdefghijk'
 * );
 * protected $redis_host = 'localhost';
 * protected $redis_port = 6379;
 * protected $namespace = 'project_name';
 * protected $increaseStepSize = 1;
 *
 * // @var JSClientHandler|null
 * protected $handlerInstance = null;
 * public function __construct()
 * {
 * $this->handlerInstance = new JSClientHandler($this->redis_host, $this->redis_port, $this->namespace, $this->token, $this->increaseStepSize);
 * }
 *
 * // 读取
 * public function getAction()
 * {
 * $this->handlerInstance->handleGet($_REQUEST['token'], $_REQUEST['group'], $_REQUEST['keys'], $_REQUEST['callback']);
 * }
 *
 * // 递增
 * public function increaseAction()
 * {
 * $this->handlerInstance->handleIncrease($_REQUEST['token'], $_REQUEST['group'], $_REQUEST['keys'], $_REQUEST['callback']);
 * }
 *
 * // 排名
 * public function rankAction()
 * {
 * $this->handlerInstance->handleRank($_REQUEST['token'], $_REQUEST['group'], $_REQUEST['type'], $_REQUEST['limit'], $_REQUEST['callback']);
 * }
 *
 * // 最高的十个数据
 * public function top10Action()
 * {
 * $this->handlerInstance->handleTop10($_REQUEST['token'], $_REQUEST['group'], $_REQUEST['callback']);
 *
 * }
 *
 * // 最低的十个数据
 * public function down10Action()
 * {
 * $this->handlerInstance->handleDown10($_REQUEST['token'], $_REQUEST['group'], $_REQUEST['callback']);
 *
 * }
 * }
 * ```
 */

class JSClientHandler
{
    /**
     * @var array 供验证的 token
     * @see JSClientHandler::__construct()
     */
    private $tokens = array();
    /**
     * @var string 命名空间
     */
    private $namespace = null;
    /**
     * @var int 递增的步长
     */
    private $increaseStepSize = 1;
    /**
     * @var CounterRank
     */
    private $counterRank = NULL;
    /**
     * @var string 多个键间的分隔符
     */
    private $keysSeparator = '_';
    /**
     * @var bool
     */
    private $isOutput = true;
    /**
     * @var string 最后一次输出的数据
     */
    private $lastOutputData = '';
    /**
     * @var string 回调函数验证正则，防止 XSS
     */
    private $callbackNameRegex = '/^[a-zA-z\$_][\w_\$]*$/';

    /**
     *
     *
     * @var \Closure
     */
    private $tokenVerifier = null;

    /**
     * @param string $redis_host redis host
     * @param int $redis_port redis port
     * @param string $namespace 顶级命名空间
     * @param array $tokens 用于与客户端访问权限控制，格式是  array('groupName'=>'token', 'groupName1'=>'token1', 'groupNameN'=>'tokenN')
     * @param int $increaseStepSize 递增步长
     * @param string $keysSeparator 多个键间的分隔符，默认是 _
     * @param bool $useFloat 是否使用浮点数，默认不使用
     * @param bool $isOutput 是否输出，默认输出。不管是否输出，都可以通过 getLastOutput() 获取最后一次的输出数据
     */
    public function __construct($redis_host, $redis_port, $namespace, array $tokens, $increaseStepSize = 1, $keysSeparator = '_', $useFloat = false, $isOutput = true)
    {
        $this->tokens = $tokens;
        $this->namespace = trim($namespace);
        $this->increaseStepSize = $increaseStepSize;
        $this->keysSeparator = $keysSeparator;
        $this->isOutput = $isOutput;

        $this->counterRank = new CounterRank($redis_host, $redis_port, $namespace, '', $useFloat);
    }

    /**
     * 获取当前使用的 CounterRank 对象
     *
     * @return CounterRank
     */
    public function getCounterRankInstance()
    {
        return $this->counterRank;
    }

    /**
     * 处理 get 操作
     * 路径设计参考：/client.js/:token/:group/:item/get?callback=:callback
     *
     * @param string $userHash 客户端用户提交的 token
     * @param string $userGroupName 客户端用户提交的分组名
     * @param string $userKeys 客户端用户提交的 item key
     * @param string $userCallback 客户端用户提交的 JsonP 回调函数，可留空，留空则直接输出值
     */
    public function handleGet($userHash, $userGroupName, $userKeys, $userCallback = '')
    {
        $this->baseInfoCheck($userHash, $userGroupName, 'get', $userKeys);


        $keys = explode($this->keysSeparator, $userKeys);
        if (is_array($keys) && count($keys) > 1) {
            $items = $this->counterRank->mGet($keys);
        } else {
            $items = $this->counterRank->get($userKeys);
        }

        $this->outputResult($items, $userCallback);
    }

    /**
     * 递增处理
     * URL 设计参考： /client.js/:token/:group/:item/increase?callback=:callback
     *
     * @param string $userHash
     * @param string $userGroupName
     * @param string $userKeys
     * @param string $userCallback
     */
    public function handleIncrease($userHash, $userGroupName, $userKeys, $userCallback = '')
    {
        $this->baseInfoCheck($userHash, $userGroupName, 'increase', $userKeys);

        $keys = explode($this->keysSeparator, $userKeys);

        if (is_array($keys) && count($keys) > 1) {
            $items = $this->counterRank->mIncrease($keys, $this->increaseStepSize);
        } else {
            $items = $this->counterRank->increase($keys[0], $this->increaseStepSize);
        }

        $this->outputResult($items, $userCallback);

    }

    /**
     * 排名处理
     * 路径设计参考：/client.js/:token/:group/rank/:limit?type=:type&callback=:callback
     *
     * @param $userHash
     * @param $userGroupName
     * @param $userType
     * @param $userLimit
     * @param $userCallback
     */
    public function handleRank($userHash, $userGroupName, $userType, $userLimit, $userCallback = '')
    {
        $this->baseInfoCheck($userHash, $userGroupName, 'rank');

        $this->outputResult($this->counterRank->rank($userLimit, $userType), $userCallback);
    }

    /**
     * top10 handler
     * 路径设计参考：/client.js/:token/:group/top10?callback=:callback
     *
     * @param string $userHash
     * @param string $userGroupName
     * @param string $userCallback
     */
    public function handleTop10($userHash, $userGroupName, $userCallback = '')
    {
        $this->handleRank($userHash, $userGroupName, 'desc', 10, $userCallback);
    }

    /**
     * down 10 handler
     * 路径设计参考：/client.js/:token/:group/down10?callback=:callback
     *
     * @param string $userHash
     * @param string $userGroupName
     * @param string $userCallback
     */
    public function handleDown10($userHash, $userGroupName, $userCallback)
    {
        $this->handleRank($userHash, $userGroupName, 'asc', 10, $userCallback);
    }

    /**
     * 基础信息检查：token 验证，设置分组名等等
     *
     * @param string $userHash
     * @param string $userGroupName
     * @param string $operation
     * @param string $keys
     */
    private function baseInfoCheck($userHash, $userGroupName, $operation, $keys = null)
    {
        if (!$userHash) {
            $this->outputError('token 未指定');
        }
        if (!isset($this->tokens[$userGroupName]) || !$this->verifyToken($operation, $userHash, $this->tokens[$userGroupName], $userGroupName, $keys)) {
            $this->outputError('token 不正确');
        }
        $userGroupName = trim($userGroupName);

        if (!$userGroupName) {
            $this->outputError('分组名未指定');
        }

        $this->counterRank->setGroupName($userGroupName);

    }

    /**
     * 输出结果
     *
     * @param array|string $items
     * @param string $userCallback
     */
    private function outputResult($items, $userCallback = '')
    {
        if (is_array($items)) {
            $items = json_encode($items);
        }
        if ($items === false || $items === null) {
            $items = 'null';
        }
        if ($userCallback) {
            if (!preg_match($this->callbackNameRegex, $userCallback)) {
                $this->outputError('invalid callback name.');
            }
            $this->output("{$userCallback}({$items});");
        } else {
            $this->output($items);
        }
    }

    /**
     * 输出信息
     *
     * @param $string
     */
    private function output($string)
    {
        if ($this->isOutput) {
            echo $string;
        }
        $this->lastOutputData = $string;
    }

    /**
     * 获取最后一次的输出
     *
     * @return string
     */
    public function getLastOutput()
    {
        return $this->lastOutputData;
    }

    /**
     * 输出错误信息，并结束后续执行
     *
     * @param $message
     */
    private function outputError($message)
    {
        die("throw new Error('{$message}');");
    }

    /**
     * token 验证
     *
     * @param string $operation 当前执行的操作
     * @param string $userToken 用户提交的 token
     * @param string $token 当前 group 的token
     * @param string $group 当前分组名
     * @param string $key 当前操作的 $key
     *
     * @return bool
     */
    protected function verifyToken($operation, $userToken, $token, $group, $key)
    {
        if ($this->tokenVerifier instanceof \Closure) {
            return call_user_func_array($this->tokenVerifier, func_get_args());
        } else {
            return $userToken === $token;
        }
    }

    /**
     * 设置 token 验证器，该验证器接收的参数与 verifyToken 相同，验证通过返回 true ，否则返回 false。
     * @see JSClientHandler::verifyToke()
     *
     * @param callable $tokenVerifier
     */
    public function setTokenVerifier(\Closure $tokenVerifier)
    {
        $this->tokenVerifier = $tokenVerifier;
    }

    /**
     * 移除 token 验证器
     */
    public function removeTokenVerifier()
    {
        $this->tokenVerifier = null;
    }
}